File: src/dsa/handbook/SelectionSort.java
package dsa.handbook;

/*
==================================================
📌 PROBLEM: Selection Sort
==================================================

Selection sort fixes one position per pass.
For index i, it selects minimum from suffix i..n-1 and places it at i.

You are given:
- unsorted array

👉 Your task:
sort in ascending order by repeated minimum selection.

--------------------------------------------------
🧠 1. INTUITION (DEEP UNDERSTANDING)
--------------------------------------------------

Invariant after pass i:
prefix 0..i is sorted and contains smallest i+1 values.

Each pass:
scan suffix for minIdx, then perform one swap.

--------------------------------------------------
❌ 2. WRONG APPROACH (BEGINNER MISTAKE)
--------------------------------------------------

Mistake:
swap every time a smaller value is found during scan.

Why weak:
it still sorts but does unnecessary writes.
Selection sort is valued for at most one swap per pass.

--------------------------------------------------
🧪 3. STEP-BY-STEP DRY RUN (REAL EXAMPLE)
--------------------------------------------------

arr=[64,25,12,22,11]
i=0 -> min=11 at idx4 -> [11,25,12,22,64]
i=1 -> min=12 at idx2 -> [11,12,25,22,64]
i=2 -> min=22 at idx3 -> [11,12,22,25,64]

--------------------------------------------------
📊 4. VISUAL REPRESENTATION
--------------------------------------------------

Visual split:
left side sorted prefix, right side unsorted suffix.

--------------------------------------------------
⚙️ 5. APPROACHES
--------------------------------------------------

Selection sort complexity:
- Time: O(n^2) in all cases
- Space: O(1)

--------------------------------------------------
⚠️ 6. EDGE CASES
--------------------------------------------------

- empty array
- single element
- duplicates
- already sorted input

--------------------------------------------------
❌ 7. COMMON MISTAKES
--------------------------------------------------

- not resetting minIdx at each outer loop
- looping j from 0 instead of i+1
- unnecessary self-swap confusion (harmless but noisy)

--------------------------------------------------
🎯 8. INTERVIEW INSIGHT
--------------------------------------------------

Interview insight:
Selection sort has fewer swaps than bubble sort,
which can matter when write cost is high.

--------------------------------------------------
🧾 9. CLEAN CODE
--------------------------------------------------
- Keep scan and swap logic compact
- Use descriptive min index variable
- Keep demonstration input small and readable

--------------------------------------------------
🧪 10. MAIN METHOD TESTING
--------------------------------------------------
Format:
Input -> Expected -> Actual

--------------------------------------------------
📌 11. MINI SUMMARY
--------------------------------------------------

Selection sort is simple:
find minimum in remaining part and place it once.

==================================================
*/
import java.util.*;

/*
Problem: Selection Sort
Repeatedly place minimum element at current index.
*/
public class SelectionSort {
    // Standard selection sort (brute/optimal for this strategy).
    // Time: O(n^2), Space: O(1)
    public static void selectionSort(int[] a) {
        for (int i = 0; i < a.length - 1; i++) {
            int minIdx = i;
            for (int j = i + 1; j < a.length; j++) {
                if (a[j] < a[minIdx]) minIdx = j;
            }
            int t = a[i]; a[i] = a[minIdx]; a[minIdx] = t;
        }
    }

    public static void main(String[] args) {
        int[] a = {64,25,12,22,11};

        selectionSort(a);
        System.out.println("Input: [64, 25, 12, 22, 11]");
        System.out.println("Expected: [11, 12, 22, 25, 64] | Actual: " + Arrays.toString(a));
    }
}

File: visualizer-platform/algorithms/selectionSort.js
import { parseNumberList } from "./helpers.js";

const DEFAULT_ARRAY = [29, 10, 14, 37, 13, 5, 22];

export const selectionSortAlgorithm = {
  key: "selectionSort",
  label: "Selection Sort",
  category: "array",
  description: "Repeatedly selects the minimum from the unsorted suffix.",
  codeLines: [
    "for (int i = 0; i < n - 1; i++) {",
    "  int min = i;",
    "  for (int j = i + 1; j < n; j++) {",
    "    if (a[j] < a[min]) min = j;",
    "  }",
    "  swap(a[i], a[min]);",
    "}"
  ],
  inputSchema: [
    { id: "arrayText", label: "Array", type: "text", defaultValue: DEFAULT_ARRAY.join(",") }
  ],
  createDefaultInput() {
    return { arrayText: DEFAULT_ARRAY.join(",") };
  },
  buildTimeline(input) {
    const a = parseNumberList(input.arrayText, DEFAULT_ARRAY);
    const steps = [];
    let comparisons = 0;
    let swaps = 0;

    steps.push(makeStep(a, { i: 0, j: "-", min: 0, comparisons, swaps }, "Start selection sort", "Each pass places one element in final position.", [], [], [], 1));

    for (let i = 0; i < a.length - 1; i += 1) {
      let min = i;
      for (let j = i + 1; j < a.length; j += 1) {
        comparisons += 1;
        steps.push(makeStep(a, { i, j, min, comparisons, swaps }, `Compare index ${j} with current min ${min}`, "Find smallest element in unsorted region.", [j, min], [], range(0, i), 3));
        if (a[j] < a[min]) {
          min = j;
          steps.push(makeStep(a, { i, j, min, comparisons, swaps }, `Update min index to ${min}`, "New minimum candidate found.", [j, min], [], range(0, i), 4));
        }
      }
      [a[i], a[min]] = [a[min], a[i]];
      swaps += 1;
      steps.push(makeStep(a, { i, j: "-", min, comparisons, swaps }, `Swap index ${i} and ${min}`, "Place selected minimum at front of unsorted region.", [i, min], [i, min], range(0, i + 1), 6));
    }

    steps.push(makeStep(a, { i: a.length - 1, j: "-", min: a.length - 1, comparisons, swaps }, "Array sorted", "All positions are finalized.", [], [], range(0, a.length), 7));
    return { timeline: steps, trackerKeys: ["i", "j", "min", "comparisons", "swaps"] };
  }
};

function makeStep(values, tracker, description, why, compare, swapped, sorted, codeLine) {
  return {
    description,
    why,
    stateSnapshot: { mode: "array", values: [...values] },
    highlightedElements: { compare, swapped, sorted },
    codeLine,
    tracker
  };
}

function range(start, end) {
  const out = [];
  for (let i = start; i < end; i += 1) out.push(i);
  return out;
}
